### PR TITLE
[FLINK-14579][sql cli] enable SQL CLI to configure modules via yaml config

### PR DESCRIPTION
**Pull request by bowenli86**
### What is the purpose of the change

enable SQL CLI to configure modules via yaml config

## Verifying this change

This change added tests and can be verified in `ExecutionContextTest` and `DependencyTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

docs will be added later

### CODE REVIEW

**Comment by bowenli86** [comment]
@xuefuz @lirui-apache @zjuwangg

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java [hunk 1]
```diff
@@ -74,6 +78,24 @@ public Environment() {
 		this.deployment = DeploymentEntry.DEFAULT_INSTANCE;
 	}
 
+	public Map<String, ModuleEntry> getModules() {
+		return modules;
+	}
+
+	public void setModules(List<Map<String, Object>> modules) {
```
Comment (xuefuz) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java]: Can we rename the input arg so as to make the code cleaner?

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java [hunk 2]
```diff
@@ -74,6 +78,24 @@ public Environment() {
...
+		this.modules = new HashMap<>(modules.size());
+
+		modules.forEach(config -> {
+			final ModuleEntry module = ModuleEntry.create(config);
+			if (this.modules.containsKey(module.getName())) {
+				throw new SqlClientException(
+					String.format("Cannot create module '%s' because a module with this name is already registered.",
```
Comment (xuefuz) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java]: registered => created.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java [hunk 3]
```diff
@@ -74,6 +78,24 @@ public Environment() {
...
 
+	public Map<String, ModuleEntry> getModules() {
+		return modules;
+	}
+
+	public void setModules(List<Map<String, Object>> modules) {
+		this.modules = new HashMap<>(modules.size());
+
+		modules.forEach(config -> {
+			final ModuleEntry module = ModuleEntry.create(config);
```
Comment (xuefuz) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java]: module => moduleEntry or entry?

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java [hunk 4]
```diff
@@ -153,6 +175,11 @@ public DeploymentEntry getDeployment() {
 	@Override
 	public String toString() {
 		final StringBuilder sb = new StringBuilder();
+		sb.append("===================== Modules =====================\n");
+		modules.forEach((name, catalog) -> {
```
Comment (xuefuz) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java]: catalog?

File: flink-table/flink-sql-client/src/test/resources/test-sql-client-modules.yaml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+deployment:
+  response-timeout: 5000
+
+modules:
+  - name: core
+    type: core
+  - name: mymodule
+    type: ModuleDependencyTest
+    test: test
```
Comment (xuefuz) [inline flink-table/flink-sql-client/src/test/resources/test-sql-client-modules.yaml]: What is this for?

File: flink-table/flink-sql-client/conf/sql-client-defaults.yaml [hunk 6]
```diff
@@ -78,6 +78,15 @@ catalogs: [] # empty list
 #    hive-conf-dir: /opt/hive_conf/
 #    default-database: ...
 
+#==============================================================================
+# Modules
+#==============================================================================
+
+# Define modules here.
+
+#modules: # note the following modules will be of the order they are specified
```
Comment (zjuwangg) [inline flink-table/flink-sql-client/conf/sql-client-defaults.yaml]: should here be `modules: [] # empty list `
like above catalogs

Comment (bowenli86) [reply flink-table/flink-sql-client/conf/sql-client-defaults.yaml]: not really. that would wipe out the default core module

**Comment by bowenli86** [comment]
@xuefuz @zjuwangg updated, pls take another look

File: flink-table/flink-sql-client/src/test/resources/test-sql-client-modules.yaml (hunk 5 above)
Comment (bowenli86) [reply flink-table/flink-sql-client/src/test/resources/test-sql-client-modules.yaml]: to test customized modules can be loaded, similar to test of catalogs

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java (hunk 1 above)
Comment (xuefuz) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java]: Did you address this comment?

Comment (bowenli86) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java]: it conforms to naming in other APIs of the class, so we should be good

File: flink-table/flink-sql-client/conf/sql-client-defaults.yaml (hunk 6 above)
Comment (zjuwangg) [reply flink-table/flink-sql-client/conf/sql-client-defaults.yaml]: ok

**Comment by bowenli86** [comment]
@xuefuz can you take another look?

**Comment by bowenli86** [comment]
@xuefuz thanks! merging

**Pull request closed without merging**